### PR TITLE
bpf: Tests for session affinity and bpf-lb-sock-hostns-only

### PR DESCRIPTION
**Pull request by dylandreimerink**
Has tests for (https://github.com/cilium/cilium/issues/20160):
* Tests NodePort with sessionAffinity
* Checks connectivity when skipping socket lb in pod ns

Adopted from #20423

### CODE REVIEW

File: bpf/tests/session_affinity_test.c [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+{
+	struct pktgen builder;
+	struct ethhdr *eh;
+	struct iphdr *iph;
+	struct tcphdr *tcph;
+
+	pktgen__init(&builder, ctx);
+
+	eh = pktgen__push_ethhdr(&builder);
+	if (!eh) return TEST_ERROR;
```
Comment (julianwiedmann) [inline bpf/tests/session_affinity_test.c]: All these same-line `return` statements make my brain hurt :).

File: bpf/lib/lb.h [hunk 2]
```diff
@@ -1271,6 +1271,18 @@ lb4_select_backend_id(struct __ctx_buff *ctx __maybe_unused,
 	index = hash_from_tuple_v4(tuple) % LB_MAGLEV_LUT_SIZE;
         return map_array_get_32(backend_ids, index, (LB_MAGLEV_LUT_SIZE - 1) << 2);
 }
+#elif LB_SELECTION == LB_SELECTION_FIRST
+/* Backend selection for tests that always chooses first slot. */
+static __always_inline __u32
+lb4_select_backend_id(struct __ctx_buff *ctx,
```
Comment (julianwiedmann) [inline bpf/lib/lb.h:1289]: Think we should also add an IPv6 variant.

File: bpf/include/bpf/ctx/xdp.h [hunk 3]
```diff
@@ -309,8 +309,19 @@ ctx_redirect_peer(const struct xdp_md *ctx __maybe_unused,
 static __always_inline __maybe_unused __u64
 ctx_full_len(const struct xdp_md *ctx)
 {
-	/* No non-linear section in XDP. */
-	return ctx_data_end(ctx) - ctx_data(ctx);
+	__u64 len;
+	/* Compute the length using inline assembly as clang
+	 * sometimes reorganizes expressions involving this,
+	 * which leads to "pointer arithmetict on pkt_end prohibited"
```
Comment (julianwiedmann) [inline bpf/include/bpf/ctx/xdp.h]:
```suggestion
	 * which leads to "pointer arithmetic on pkt_end prohibited"
```

File: bpf/tests/session_affinity_test.c (hunk 1 above)
Comment (dylandreimerink) [reply bpf/tests/session_affinity_test.c]: Agreed, we can move the returns to the next line

**Comment by julianwiedmann** [comment]
>  bpf: Add BPF unit test for bpf-lb-sock-hostns-only
> 
> Tests that the socket lb is skipped if ENABLE_SOCKET_LB_HOST_ONLY
> is set and the packet is **not** from host namespace.

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Looks good, ty! But please do add the `lb6_select_backend_id()` for `LB_SELECTION_FIRST`.

**Review by nebril — APPROVED** [review]
`test` bits LGTM

**Review by julianwiedmann — APPROVED** [review]
Thank you!

**Comment by dylandreimerink** [comment]
Both the `k8s-1.16-kernel-4.9 (test-1.16-4.9)` and `Cilium L4LB XDP (ci-l4lb)` appear to be flakes since these same tests are failing on other branches as well, the rest of the tests are all green.

**Pull request merged**